modules/tinymce/src/core/main/ts/api/html/DomParser.ts

```diff
@@ -90,276 +92,263 @@ interface WalkResult {
+      Replication.mutate(element, rule.outputName as keyof HTMLElementTagNameMap);
+    }
+  });
+
+  // Let's do the same thing for attributes
+  purify.addHook('uponSanitizeAttribute', (ele, evt) => {
+    evt.keepAttr = evt.attrName.startsWith('data-') || schema.isValid(ele.tagName.toLowerCase(), evt.attrName);
+    if (evt.keepAttr) {
+      if (!Obj.has(evt.allowedAttributes, evt.attrName)) {
+        evt.allowedAttributes[evt.attrName] = true;
       }
 
-      // Do we need to create the node
-      if (text.length !== 0) {
-        textNode = new AstNode('#text', 3);
-        textNode.raw = !!raw;
-        node.append(textNode).value = text;
+      if (evt.attrName in schema.getBoolAttrs()) {
+        evt.attrValue = evt.attrName;
       }
-    },
+    }
+  });
 
-    comment: (text) => {
-      node.append(new AstNode('#comment', 8)).value = text;
-    },
+  return purify;
+};
 
-    pi: (name, text) => {
-      node.append(new AstNode(name, 7)).value = text;
-      removeWhitespaceBefore(node);
-    },
+const transferChildren = (parent: AstNode, nativeParent: Node) => {
+  const isSpecial = specialElements.has(parent.name);
+  Arr.each(nativeParent.childNodes, (nativeChild) => {
+    const child = new AstNode(nativeChild.nodeName.toLowerCase(), nativeChild.nodeType);
+
+    if (NodeType.isElement(nativeChild)) {
+      Arr.each(nativeChild.attributes, (attr) => {
+        child.attr(attr.name, attr.value);
+      });
+    } else if (NodeType.isText(nativeChild)) {
+      child.value = nativeChild.data;
+      if (isSpecial) {
+        child.raw = true;
+      }
+    } else if (NodeType.isComment(nativeChild) || NodeType.isCData(nativeChild) || NodeType.isPi(nativeChild)) {
+      child.value = nativeChild.data;
+    }
 
-    doctype: (text) => {
-      const newNode = node.append(new AstNode('#doctype', 10));
-      newNode.value = text;
-      removeWhitespaceBefore(node);
-    },
+    transferChildren(child, nativeChild);
+    parent.append(child);
+  });
+};
 
-    start: (name, attrs, empty) => {
-      const newNode = new AstNode(name, 1);
-      newNode.attributes = attrs;
+const walker = (root: AstNode, schema: Schema, nodeFilters: Record<string, ParserFilterCallback[]>, attributeFilters: ParserFilter[]): WalkResult => {
+  const state = { invalidChildren: [], matchedNodes: {}, matchedAttributes: {}};
 
-      node.append(newNode);
+  let node = root;
+  while ((node = node.walk())) {
+    filterNode(node, nodeFilters, attributeFilters, state);
 
-      // Trim whitespace before block
-      if (blockElements[name]) {
-        removeWhitespaceBefore(newNode);
-      }
+    const parent = node.parent;
+    if (!parent) {
+      continue;
+    }
 
-      // Change current node if the element wasn't empty i.e not <br /> or <img />
-      if (!empty) {
-        node = newNode;
-      }
+    // Check if node is valid child of the parent node is the child is
+    // unknown we don't collect it since it's probably a custom element
```

Comment:
I assume this is more correct?
```suggestion
    // Check if the node is a valid child of the parent node. If the child is
    // unknown we don't collect it since it's probably a custom element
```